lib/testing/src/equality/equality.dart

```diff
@@ -0,0 +1,229 @@
+        .add(" ${matchState[FAILURE_REASON]}");
+
+
+  void _verifyEqualityGroups(Map<String, List<Object>> equalityGroups,
+      Map matchState) {
+    if (equalityGroups == null) {
+      throw new MatchError('Equality Group must not be null');
+    }
+    var equalityGroupsCopy = {};
+    equalityGroups.keys.forEach((String groupName) {
+      if (groupName == null) {
+        throw new MatchError('Group name must not be null');
+      }
+      var group = equalityGroups[groupName];
+      if (group == null) {
+        throw new MatchError('Group must not be null');
+      }
+      equalityGroupsCopy[groupName] = new List.from(group);
+    });
+
+    // Run the test multiple times to ensure deterministic equals
+    for (var run in range(REPETITIONS)) {
+      _checkBasicIdentity(equalityGroupsCopy, matchState);
+      _checkGroupBasedEquality(equalityGroupsCopy);
+    }
+  }
+
+  void _checkBasicIdentity(Map<String, List<Object>> equalityGroups,
+      Map matchState) {
+    var flattened = equalityGroups.values.expand((group) => group);
+    for (var item in flattened) {
+      if (item == _NotAnInstance.EQUAL_TO_NOTHING) {
+        throw new MatchError("$item must not be Object#equals to an arbitrary "
+            "object of another class");
+      }
+
+      if (item != item) {
+        throw new MatchError("$item must be Object#equals to itself");
+      }
+
+      var a = item.hashCode;
+      var b = item.hashCode;
+      if (a != b) {
+        throw new MatchError("the Object#hashCode of $item must be consistent");
+      }
+    }
+  }
+
+  void _checkGroupBasedEquality(Map<String, List<Object>> equalityGroups) {
+    equalityGroups.keys.forEach((String groupName) {
+      var groupLength = equalityGroups[groupName].length;
+      for (var itemNumber = 0; itemNumber < groupLength; itemNumber++) {
+        _checkEqualAgainstSameGroup(equalityGroups, groupLength, itemNumber,
+            groupName);
+        _checkUnequalsAgainstOtherGroups(equalityGroups, groupName, itemNumber);
+      }
+    });
+  }
+
+  void _checkUnequalsAgainstOtherGroups(Map<String,
+      List<Object>> equalityGroups, String groupName, int itemNumber) {
+    equalityGroups.keys.forEach((String unrelatedGroupName) {
+      if (groupName != unrelatedGroupName) {
+        var unrelatedGroup = equalityGroups[unrelatedGroupName];
+        for (var unrelatedItemNumber = 0;
+            unrelatedItemNumber < unrelatedGroup.length;
+            unrelatedItemNumber++) {
+          _expectUnrelated(
+              equalityGroups,
+              groupName,
+              itemNumber,
+              unrelatedGroupName,
+              unrelatedItemNumber);
+        }
+      }
+    });
+  }
+
+  void _checkEqualAgainstSameGroup(Map<String, List<Object>> equalityGroups,
+    int groupLength, int itemNumber, String groupName) {
+    for (var relatedItemNumber = 0; relatedItemNumber < groupLength;
+        relatedItemNumber++) {
+      if (itemNumber != relatedItemNumber) {
```

Comment:
I think the implementation might be simpler if checking within the same group included checking self equality. If equality check fails, then check the index to issue different error messages.